saleor/payment/providers/braintree/__init__.py

```diff
@@ -0,0 +1,258 @@
+    transaction_token = gateway.client_token.generate()
+    return transaction_token
+
+
+def authorize(
+        payment: Payment,
+        transaction_token: str,
+        **connection_params: Dict) -> Tuple[Transaction, str]:
+    gateway = get_gateway(**connection_params)
+    try:
+        result = gateway.transaction.sale({
+            'amount': str(payment.total),
+            'payment_method_nonce': transaction_token,
+            'options': {
+                'submit_for_settlement': CONFIRM_MANUALLY,
+                'three_d_secure': {
+                    'required': THREE_D_SECURE_REQUIRED}},
+            **get_customer_data(payment)})
+    except braintree_sdk.exceptions.NotFoundError:
+        return transaction_and_incorrect_token_error(
+            payment, type=TransactionType.AUTH, token=transaction_token)
+    gateway_response = extract_gateway_response(result)
+    error = get_error_for_client(gateway_response['errors'])
+    txn = create_transaction(
+        payment=payment,
+        transaction_type=TransactionType.AUTH,
+        amount=gateway_response.pop('amount', payment.total),
+        currency=gateway_response.pop('currency', payment.currency),
+        gateway_response=gateway_response,
+        token=getattr(result.transaction, 'id', ''),
+        is_success=result.is_success)
+    return txn, error
+
+
+def capture(
+        payment: Payment,
+        amount: Decimal,
+        **connection_params: Dict) -> Tuple[Transaction, str]:
+    gateway = get_gateway(**connection_params)
+    auth_transaction = payment.transactions.filter(
+        transaction_type=TransactionType.AUTH, is_success=True).first()
+    try:
+        result = gateway.transaction.submit_for_settlement(
+            transaction_id=auth_transaction.token, amount=str(amount))
+    except braintree_sdk.exceptions.NotFoundError:
+        return transaction_and_incorrect_token_error(
+            payment,
+            type=TransactionType.CAPTURE,
+            token=auth_transaction.token,
+            amount=amount)
+
+    gateway_response = extract_gateway_response(result)
+    error = get_error_for_client(gateway_response['errors'])
+
+    txn = create_transaction(
+        payment=payment,
+        transaction_type=TransactionType.CAPTURE,
+        amount=gateway_response.pop('amount', amount),
+        currency=gateway_response.pop('currency', payment.currency),
+        token=getattr(result.transaction, 'id', ''),
+        is_success=result.is_success,
+        gateway_response=gateway_response)
+    return txn, error
+
+
+def void(
+        payment: Payment,
+        **connection_params: Dict) -> Tuple[Transaction, str]:
+    gateway = get_gateway(**connection_params)
+    auth_transaction = payment.transactions.filter(
+        transaction_type=TransactionType.AUTH, is_success=True).first()
+    try:
+        result = gateway.transaction.void(
+            transaction_id=auth_transaction.token)
```

Comment:
How to reproduce this error:
1. Fail a payment in braintree (get the `Unable to process transaction. Please try again in a moment` error):
2. Go to the order details;
3. Click on `Payment details`;
4. Click the big button `I understand, allow me to pay again`;
5. The error should occur.

```
ERROR django.request Internal Server Error: /en/order/f6b4d02a-3013-430d-8bf9-0281b0126343/cancel-payment/ [PID:10962:Thread-244]
Traceback (most recent call last):
  File "/develop/virtualenvs/dummy-saleor-env/lib/python3.6/site-packages/django/core/handlers/exception.py", line 35, in inner
    response = get_response(request)
  File "/develop/virtualenvs/dummy-saleor-env/lib/python3.6/site-packages/django/core/handlers/base.py", line 119, in _get_response
    response = middleware_method(request, callback, callback_args, callback_kwargs)
  File "/develop/virtualenvs/dummy-saleor-env/lib/python3.6/site-packages/debug_toolbar/middleware.py", line 84, in process_view
    response = panel.process_view(request, view_func, view_args, view_kwargs)
  File "/develop/virtualenvs/dummy-saleor-env/lib/python3.6/site-packages/debug_toolbar/panels/profiling.py", line 155, in process_view
    return self.profiler.runcall(view_func, *args, **view_kwargs)
  File "/usr/lib64/python3.6/cProfile.py", line 109, in runcall
    return func(*args, **kw)
  File "/develop/saleor/saleor/order/utils.py", line 35, in decorator
    return func(*args, **kwargs)
  File "/develop/saleor/saleor/order/views.py", line 127, in cancel_payment
    form.save()
  File "/develop/saleor/saleor/order/forms.py", line 48, in save
    payment.void()
  File "/develop/saleor/saleor/payment/models.py", line 103, in void
    return utils.gateway_void(payment=self)
  File "/develop/saleor/saleor/payment/utils.py", line 61, in func
    return view(payment, *args, **kwargs)
  File "/develop/saleor/saleor/payment/utils.py", line 146, in gateway_void
    txn, error = provider.void(payment, **provider_params)
  File "/develop/saleor/saleor/payment/providers/braintree/__init__.py", line 215, in void
    transaction_id=auth_transaction.token)
AttributeError: 'NoneType' object has no attribute 'token'
ERROR django.request Internal Server Error: /en/order/f6b4d02a-3013-430d-8bf9-0281b0126343/cancel-payment/ [PID:10962:Thread-244]
Traceback (most recent call last):
  File "/develop/virtualenvs/dummy-saleor-env/lib/python3.6/site-packages/django/core/handlers/exception.py", line 35, in inner
    response = get_response(request)
  File "/develop/virtualenvs/dummy-saleor-env/lib/python3.6/site-packages/django/core/handlers/base.py", line 119, in _get_response
    response = middleware_method(request, callback, callback_args, callback_kwargs)
  File "/develop/virtualenvs/dummy-saleor-env/lib/python3.6/site-packages/debug_toolbar/middleware.py", line 84, in process_view
    response = panel.process_view(request, view_func, view_args, view_kwargs)
  File "/develop/virtualenvs/dummy-saleor-env/lib/python3.6/site-packages/debug_toolbar/panels/profiling.py", line 155, in process_view
    return self.profiler.runcall(view_func, *args, **view_kwargs)
  File "/usr/lib64/python3.6/cProfile.py", line 109, in runcall
    return func(*args, **kw)
  File "/develop/saleor/saleor/order/utils.py", line 35, in decorator
    return func(*args, **kwargs)
  File "/develop/saleor/saleor/order/views.py", line 127, in cancel_payment
    form.save()
  File "/develop/saleor/saleor/order/forms.py", line 48, in save
    payment.void()
  File "/develop/saleor/saleor/payment/models.py", line 103, in void
    return utils.gateway_void(payment=self)
  File "/develop/saleor/saleor/payment/utils.py", line 61, in func
    return view(payment, *args, **kwargs)
  File "/develop/saleor/saleor/payment/utils.py", line 146, in gateway_void
    txn, error = provider.void(payment, **provider_params)
  File "/develop/saleor/saleor/payment/providers/braintree/__init__.py", line 215, in void
    transaction_id=auth_transaction.token)
AttributeError: 'NoneType' object has no attribute 'token'
```

Reply:
It's fixed on my local instance